src/main/scala/com/yotpo/metorikku/test/Tester.scala

```diff
@@ -85,75 +91,273 @@ case class Tester(config: TesterConfig) {
-    val actualSchema = aggregatedErrorRows.head.schema
 
-    val expectedValuesRows = metricExpectedResultRows.map(expectedRes => {
-      Row.fromSeq(expectedRes.values.toSeq)
-    })
+  def compareRowsByAllCols(actualResultRow: Map[String, Any], rowIndex: Int, sortedExpectedResults: List[Map[String, Any]], tableKeys: List[String], metricName: String, tableName: String, mapSortedToExpectedIndexes: mutable.Map[Int, Int]): String = {
+    val expectedResultRow = sortedExpectedResults(rowIndex)
+    val mismatchingCols = getMismatchingColumns(actualResultRow, expectedResultRow)
+    if (mismatchingCols.length > 0) {
+      val tableKeysVal = getRowKey(expectedResultRow, tableKeys)
+      val outputKey = formatOutputKey(tableKeysVal, tableKeys)
+      return s"[$metricName - $tableName] failed on original row ${mapSortedToExpectedIndexes(rowIndex) + 1}, sorted row_id ${rowIndex + 1} with key ${outputKey}." +
+        s" Column values mismatch on ${mismatchingCols.mkString(", ")}"
+    }
+    return ""
+  }
+
+  def getSubDf(sortedExpectedRows: List[Map[String, Any]], errorsIndexArr: Seq[Int]) = {
+//    var res = List[Map[String, Any]]()
+//    for ((row, index) <- sortedExpectedRows.zipWithIndex) {
+//      if (errorsIndexArr.contains(index)) {
+//        res += row
+//      }
+//    }
+  }
+
+  private def printTableErrors(tableErrors: Array[String], sortedExpectedRows: List[Map[String, Any]], sortedActualResults: Array[Map[String, Any]], errorsIndexArr: Seq[Int]) = {
+    println("****************************  Test failed  *******************************")
+    println("**************************  Expected results  ****************************")
+    transformListMapToDf(sortedExpectedRows).show(true)
+    println("***************************  Actual results  *****************************")
+    transformListMapToDf(sortedActualResults.toList).show(true)
+    println("******************************  Errors  **********************************")
+    for (error <- tableErrors) {
+      println(error)
+    }
+  //  println("*********************  Expected missing results  *************************")
+  //  getSubDf(sortedExpectedRows, errorsIndexArr).show()
+  //  println("*********************  Actual enexpected results  *************************")
+  //  getSubDf(sortedActualResults, errorsIndexArr).show()
 
-    val expectedSchema = getExpectedSchema(metricExpectedResultRows.head.keys, actualSchema)
-    val expectedArrayStructFields = getExpectedArrayFields(metricExpectedResultRows.head.keys, actualSchema)
+  }
 
-    val actualRowsDF = spark.sqlContext.createDataFrame(spark.sparkContext.parallelize(aggregatedErrorRows), actualSchema)
-    val expectedRowsDF = spark.sqlContext.createDataFrame(spark.sparkContext.parallelize(expectedValuesRows), expectedSchema)
-    val columns = expectedSchema.fields.map(_.name).filter(x => !expectedArrayStructFields.contains(x))
+  private def sortRows(a: Map[String, Any], b: Map[String, Any]): Boolean = {
```

Comment:
Handle subsets of same key in a separate function, it's making everything more complicated.